### ISSUE TITLE
EDGAR discontinuing FTP support

### ISSUE THREAD

**Issue by kevinyso** [pydata/pandas-datareader#269]
Per the SEC [website](https://www.sec.gov/edgar/searchedgar/ftpusers.htm), EDGAR is discontinuing FTP support:

>Permanent FTP Shutdown Notice
>On Friday, December 30, 2016, at 11:59 pm, Eastern Time, FTP services for retrieving EDGAR filing documents will be permanently turned off. Filings will remain available via HTTPS at https://www.sec.gov/. For HTTPS EDGAR file system guidance, see Accessing EDGAR Data.

At the time of this posting, it seems like that the server is still accessible, but discontinuation of the FTP server will inevitably break EDGAR support in pandas-datareader. A rewrite for retrieval via HTTPS will probably be necessary.

**Comment by femtotrader** [comment]
Thanks @kevinyso 

Accessing EDGAR Data
https://www.sec.gov/edgar/searchedgar/accessing-edgar-data.htm

Pinging @jtkiley who contribute on this data reader code.

**Comment by westurner** [comment]
https://github.com/rahulrrixe/SEC-Edgar/blob/master/SECEdgar/crawler.py

https://github.com/rahulrrixe/SEC-Edgar/blob/master/setup.py

- ``SUPPORTED_VERSIONS = ['2.5', '2.6', '2.7', 'PyPy', ]``
- ``install_requires=['requests', 'beautifulsoup4', 'configparser',],``

Caching:

- https://github.com/reclosedev/requests-cache
- https://github.com/ionrock/cachecontrol

**Comment by jtkiley** [comment]
I'll try to get on this in the next few weeks. It looks like a lot of it will remain pretty similar, and the use of http will probably make the reliability much better than it was with the intermittent FTP issues.

**Comment by aking1012** [comment]
Has @jtkiley started hammering on this one?  If yes, in what branch on the fork.  Trying to help out someone on quantopian, and I just bumped pandas 0.13 to 0.19, so the datareader project is now a thing for me.  
  
Doesn't have a really well defined API for date related peeling, but would people be averse to having _FULL, partial quarter to date fetches working?

**Comment by aking1012** [comment]
Okay, solved.  Removed a couple of tests, and didn't even TRY to be py2.x compliant.  Any interest in a PR?  What do you want for a PR?

**Comment by jtkiley** [comment]
@aking1012: I haven't, and I appreciate your jumping in on it. I'm an academic who tries to share things I make for my research, so my time to pitch in can be somewhat uneven.

Sure, make a PR. I haven't tried it, but it seems like you're seeing success with your changes. The additional things I can spot are the following.

1. Look into removing `requests_ftp`. That's probably better done in a separate issue. It was added for this feature, but I'm not sure if anyone else is using it. I'm happy to do this once your PR is merged.
2. Expanding the tests. That's probably also better done in a separate issue, and, again, I'm happy to look at that once this is merged. Originally, the fragility of the FTP server meant that these queries had a high likelihood to fail (silently; causing Travis timeouts), so I tried to test multiple things in as few, small pulls as possible. I'm hoping that's no longer a necessary compromise.

I envy the not even trying to be 2.x compliant part. About half of my original effort was getting my fully working 3.5 version to work across the then-currently supported versions, especially 2.6.

Again, thanks for the help on this.

**Comment by aking1012** [comment]
No problem buddy.  I nuked it pretty hard.  If py2k is still a thing, let me know.  I'm just all https://www.youtube.com/watch?v=qvJIFnk0ZWM right now.  I'm also playing with a layer on top that does fetches of individual records, but the individual fetch API doesn't allow it.  I solved the timeout issue by saying "hard limit of 5 years.  just be nice" etc etc.

**Comment by aking1012** [comment]
So, everything on 3.x passes now at 96% coverage.  Need to figure out the 2.7 tests fail.

**Comment by westurner** [comment]
On Monday, April 3, 2017, Jason Kiley <notifications@github.com> wrote:

> @aking1012 <https://github.com/aking1012>: I haven't, and I appreciate
> your jumping in on it. I'm an academic who tries to share things I make for
> my research, so my time to pitch in can be somewhat uneven.
>
> Sure, make a PR. I haven't tried it, but it seems like you're seeing
> success with your changes. The additional things I can spot are the
> following.
>
>    1. Look into removing requests_ftp. That's probably better done in a
>    separate issue. It was added for this feature, but I'm not sure if anyone
>    else is using it. I'm happy to do this once your PR is merged.
>
>

>
>    1. Expanding the tests. That's probably also better done in a separate
>    issue, and, again, I'm happy to look at that once this is merged.
>    Originally, the fragility of the FTP server meant that these queries had a
>    high likelihood to fail (silently; causing Travis timeouts), so I tried to
>    test multiple things in as few, small pulls as possible. I'm hoping that's
>    no longer a necessary compromise.
>
>
Mocking the HTTP endpoints with e.g. responses might be a good idea:
https://github.com/getsentry/responses

You can record the requests/responses with vcrpy:
https://pypi.python.org/pypi/vcrpy

https://docs.travis-ci.com/user/environment-variables/

http://stackoverflow.com/questions/12983137/how-do-detect-if-travis-ci-or-not/12989131#12989131
- CI=true
- TRAVIS=true

**Comment by bashtage** [comment]
EDGAR has been deprecated.

**Comment by westurner** [comment]
@aking1012

Is this working?

https://github.com/aking1012/pandas-datareader/blob/master/pandas_datareader/tests/test_edgar.py

https://github.com/aking1012/pandas-datareader/blob/master/pandas_datareader/edgar.py

@bashtage
Is Py2.7 support a project requirement? (I see refs to 2.7 in tox.ini and .travis.yml)

**Comment by bashtage** [comment]
None of the edgar tests work since they altered the API.  There was an attempt to rebuild it against the new API but it changed again and so the new attempt also fails now. 

Py2.7 is supported but not required.

**Comment by westurner** [comment]
@aking1012 seems to have implemented the new HTTP API in his edgar.py? Is that what you're referring to?

**Comment by bashtage** [comment]
Yes, I have a branch where I rebased this onto master.  The URLs in @aking1012 's PR were are all 404.
